crates/noirc_frontend/src/graph/mod.rs

```diff
@@ -117,7 +123,32 @@ impl Dependency {
 }
 
 impl CrateGraph {
+    pub fn root_crate_id(&self) -> &CrateId {
+        self.arena
+            .keys()
+            .find(|crate_id| crate_id.is_root())
+            .expect("Something about no crates in graph")
+    }
+
     pub fn add_crate_root(&mut self, file_id: FileId) -> CrateId {
+        for (crate_id, crate_data) in self.arena.iter() {
+            if crate_id.is_root() {
+                panic!("Cannot add two crate roots to a graph. Use `add_crate` instead.");
+            }
+
+            if crate_data.root_file_id == file_id {
+                return *crate_id;
```

Comment:
If I add FileId(1) as non-crate root, then I attempt to add FileId(1) as a crate root. Then I call `root_crate_id` should that return the crate assosciated with FileId(1)?

Reply:
psuedocode to explain:

```
let file = FileId(1);
let crate_id = self.add_crate(file);
let same_crate_id = self.add_crate_root(file);

assert_eq(crate_id, same_crate_id);

let root_crate_id = self.root_crate_id(); // Should this panic or should root_crate_id == same_crate_id
```

(I haven't ran the PR, so I might have missed something)

Reply:
I think we can add that restriction to see if it does anything (not entirely sure)

 If I try to remember back to then, my thought process was something along the lines of "If two of my dependencies use the same dependency, then they would point to the FileId and so we may insert the same FileId twice"
 
 
 To clarify, I think the above is slightly different (but your point does generalise it), because we are saying a Crate is both a normal Crate and a Root, but when we say its a crate root, it doesn't get saved as a crate root. So the following test:
 
 ```
     #[test]
    fn crate_is_normal_and_a_crate_root() {
        let file_ids = dummy_file_ids(3);

        let mut graph = CrateGraph::default();
        let crate1 = graph.add_crate(file_ids[0]);
        let crate2 = graph.add_crate_root(file_ids[0]);

        graph.root_crate_id();
    }
    ```
    
    Panics at `graph.root_crate_id();` instead of at `add_crate_root`

Reply:
Makes sense. I'll have some updates up shortly. We'll have to decide if we're okay with the panics or if we want to make these result-returning since it's a bit of a "public" API (in the sense that `noirc_driver` consumes it)